BSP-116 | Authentication service

0x103 opened the pull request:
- Okta Authn service impmementation to authenticate users.
- Okta Session service implementation to create and manage a user's session.
- Okta Authorize service implementation to manage the OAuth2.0 flow.
- Okta Token service implementation handle the OAuth2.0 tokens.
- Okta User Info service implementation to fetch a user's data based on an access token.
- Corresponding Public / Private handlers & Nats implementation.
- REST Connectors (Axios & Fetch) Interfaces.
- Tests cases ( authentication / authorization / auth code exchange )
- CircleCI configuration updated for node 14.7.3 & npm 7.20.3.
- Dockerfile.
- run file updated to remove to duplicate "test" command.

0x103 committed:
BSP-116: Authentication service.

- Okta Authn service impmementation to authenticate users.
- Okta Session service implementation to create and manage a user's session.
- Okta Authorize service implementation to manage the OAuth2.0 flow.
- Okta Token service implementation handle the OAuth2.0 tokens.
- Okta User Info service implementation to fetch a user's data based on an access token.
- Corresponding Public / Private handlers & Nats implementation.
- REST Connectors (Axios & Fetch) Interfaces.
- Tests cases ( authentication / authorization / auth code exchange )
- CircleCI configuration.
- Dockerfile.
- run file updated to remove to duplicate "test" command.

podefr commented on src/commands/authorize/authorizeCommand.ts:
```diff
@@ -0,0 +1,71 @@
+import { AuthorizeInterface } from "../../interfaces/authorize";
+
+export default class AuthorizeCommand {
```
I would strongly advise against using this pattern for creating an `AuthorizeCommand`. The class doesn't add any value compared to using an object that satisfies an interface. We need to add a lot of code (which will have to be maintained) to be able to create the object, read its state, and (de)serialize it). In other languages this boilerplate would actually be automatically generated by a preprocessor to avoid mistakes.
This pattern also introduces multiple issues:
1. if I want to serialize this class I'd have to JSON.stringify(authorizeCommand.toJSON()) instead of just JSON.stringify(authorizeCommand)
2. I can't do a `new AuthorizeCommand` without a cookie (it's optional) but with a sessionToken. Unless I pass undefined to cookie but that's not nice. We generally recommend a max number of params in a function of 2 to 3 and then pass an object for optional parameters. Now we have a lot more boilerplate code to write.
3. How would I extend this object without modifying code? If we're using a plain object I could extend my interface, but here I need to change the constructor, change the .toJSON function, add a getter etc.

podefr replied:
If we're looking for immutability (I see that all class fields are private) we can simply use `Object.freeze()` on the command. And if we're looking to validate the shape of the object, the interface will make sure that our object is valid at build time. Also, if we need to validate the object at runtime, then I'd simply suggest adding a validator function using `yup` for instance

podefr commented on src/commands/session/createSessionCommand.ts:
```diff
@@ -0,0 +1,23 @@
+import { CreateSessionInterface } from "../../interfaces/session";
+
+export default class CreateSessionCommand {
+    private readonly _sessionToken: string;
```
same here, let's just create an interface and pass an object around, like authorizationCommand

podefr commented on src/commands/token/exchangeTokenCommand.ts:
```diff
@@ -0,0 +1,53 @@
+import { ExchangeTokenInterface } from "../../interfaces/token";
+
+export default class ExchangeTokenCommand {
```
same here, let's just create an interface and pass an object around, like authorizationCommand

podefr commented on src/commands/token/refreshTokenCommand.ts:
```diff
@@ -0,0 +1,49 @@
+import { RefreshTokenInterface } from "../../interfaces/token";
+
+export default class RefreshTokenCommand {
```
same here, let's just create an interface and pass an object around, like authorizationCommand

podefr commented on src/commands/userinfo/userinfoCommand.ts:
```diff
@@ -0,0 +1,23 @@
+import { UserinfoInterface } from "../../interfaces/userinfo";
+
+export default class UserinfoCommand {
```
same here, let's just create an interface and pass an object around, like authorizationCommand

podefr commented on .circleci/config.yml:
```diff
@@ -0,0 +1,30 @@
+        executor: node/default
+        steps:
+            - checkout
+            - run: npm install --force
```
is the `--force` required given that we're always doing a fresh install?

0x103 replied:
Without the `--force`, this error appear :

`npm WARN read-shrinkwrap This version of npm is compatible with lockfileVersion@1, but package-lock.json was generated for lockfileVersion@2. I'll try to do my best with it!`
`npm ERR! Maximum call stack size exceeded
`

podefr replied:
Would work if you regenerate the lock file with a more recent version of npm?

0x103 replied:
The problem is that I am running on the latest stable version of npm which generates lockfiles on version 2 and CircleCI is on npm version 6 which doesn't support lockfiles on version 2.
Should we update the npm version each time then ?

podefr replied:
let's chat about it. I think we should maybe collectively move to npm 7 across projects.

podefr commented on Dockerfile:
```diff
@@ -0,0 +1,17 @@
+FROM node:12-alpine
```
should we try 14?

0x103 replied:
Yes, I took the version from the `tokenization-service`.

podefr commented on README.md:
```diff
@@ -1,2 +1,11 @@
+
+## How to setup Dev environment
+
+WIP
```
I'm sure you're still working on this, let's make sure we don't forget to add the basic setup instructions :)

0x103 replied:
Of course yes. I planned to write it once this PR gets accepted to that the setup is based on a default version.

podefr commented on docker-compose.yaml:
```diff
@@ -0,0 +1,18 @@
+            - "6222:6222"
+            - "8222:8222"
+    authorization-service:
+        image: jwalab/authorization-service:latest
```
let's pin this version, the latest one for auth service is 0.0.2

0x103 replied:
Updated. Thank you.

podefr commented on docker-compose.yaml:
```diff
@@ -0,0 +1,18 @@
+        ports:
+            - "8999:8999"
+    airlock:
+        image: jwalab/airlock:latest
```
let's pin this version too, the most recent airlock is 0.0.7

0x103 replied:
Updated. Thank you.

podefr commented on src/config.ts:
```diff
@@ -0,0 +1,72 @@
+    );
+}
+
+const OKTA_URI = `${PROTOCOL}://${OKTA_ORG}`;
```
should the var names have Okta in them? is the flow specific to okta or common to all oauth2 services? if the auth flow works with all oauth2 services I'd suggest removing OKTA?

0x103 replied:
Indeed, the flow is common to all OAuth2.0 services. I'll remove OKTA and also ask for the links to be defined in the environment in order to avoid any url building error.

podefr commented on src/network/fetchConnector.ts:
```diff
@@ -0,0 +1,72 @@
+import fetch from "node-fetch";
```
why do we have an axios connector and a fetch connector?

0x103 replied:
The Fetch connector has been asked as an example on how easy it is to change the default connector (axios) if needed. It works the same way for the services.

tkjwa replied:
@podefr this has been done for learning and discovering some of the new patterns introduced in the PR. I guess we could remove it before merging

0x103 committed:
BSP-116: Interfaces update.

0x103 committed:
BSP-116: exchangeTokenCommand interface updated.

0x103 committed:
BSP-116: Docker compose update.

0x103 committed:
BSP-116: Airlock tag updated.

0x103 committed:
BSP-116: Previous airlock tag restored.

0x103 committed:
BSP-116: env.

0x103 committed:
BSP-116: Shell script.

0x103 committed:
BSP-116: run script.

0x103 committed:
BSP-116: CircleCI start.

0x103 committed:
BSP-116: config fixed.

0x103 committed:
BSP-116: Class format commands & prettier.

0x103 committed:
BSP-116: Authn fix.

0x103 committed:
BSP-116: run file docker commands.

0x103 committed:
BSP-116 CircleCI Config

0x103 committed:
BSP-116: CircleCI start command updated.

podefr commented on src/private/authnPrivateHandler.ts:
```diff
@@ -0,0 +1,60 @@
+                try {
+                    const data = jsonCodec.decode(
+                        message.data
+                    ) as AuthnInterface;
```
we don't need to typecast the payload here. We directly pass it to `AuthnCommand()` and it will verify the shape of the payload and then assign it a type for us. I think AuthInterface could be removed across the code base

podefr commented on src/private/authorizePrivateHandler.ts:
```diff
@@ -0,0 +1,50 @@
+                    const { headers } = message;
+                    const data = jsonCodec.decode(
+                        message.data
+                    ) as AuthorizeInterface;
```
same here. Let's use the generic `deserialize` function to `jsonCodec.decode` payloads and assign them a type.

podefr commented on src/commands/authn/authnCommand.ts:
```diff
@@ -0,0 +1,21 @@
+    readonly password: string;
+
+    constructor(payload: AuthnInterface) {
+        const { username, password } = payload;
```
no need for the constructor to accept a typed parameter. It can just take anything that looks like a JSON, even `any` I think, and then it will validate that the parameter has the right fields and assign it a type/interface when returning. 
Also payload isn't the right name here, the constructor doesn't know where the JSON comes from, whether it's a nats payload or someone's object literal. I'd suggest a generic parameter: `data: any`

podefr commented on src/commands/token/refreshTokenCommand.ts:
```diff
@@ -0,0 +1,35 @@
+import { RefreshTokenInterface } from "../../interfaces/token";
+
+export default class RefreshTokenCommand implements RefreshTokenInterface {
```
for all of our classes, I don't think we need the implements. the class already brings all the type safety we need and the interface doesn't add any, it's not used anywhere else either. When we need to verify if a value is of type RefreshTokenCommand we just need to use the class as a type, instead of using the underlying interface.

podefr commented on src/interfaces/authn.ts:
```diff
@@ -0,0 +1,19 @@
+    };
+}
+
+interface AuthnInterface {
```
if we need to keep this one we should move it next to the Class

podefr commented on src/services/oktaSessionService.ts:
```diff
@@ -0,0 +1,28 @@
+            sessionToken: sessionToken
+        } as CreateSessionInterface;
+
+        return this._restConnector.post(SESSIONS_ENDPOINT, payload);
```
let's remove the `_`

0x103 committed:
BSP-116: Interfaces removed, commands updated, deserializer added.

0x103 committed:
BSP-116: Old dependency deleted.

0x103 committed:
BSP-116: payload var renamed, cookie correctly given to authorize private handler.

0x103 committed:
BSP-116: Unused dependency deleted.

0x103 committed:
BSP-116: Test debug.

0x103 committed:
BSP-116: Test debug.

0x103 committed:
BSP-116: Deserializer update for headers handling.

0x103 committed:
BSP-116: Deserializer headers.

0x103 committed:
BSP-116: Imports reorganized, new tests.

0x103 committed:
BSP-116: Prettier.

0x103 committed:
BSP-116: Error messages updated.

0x103 committed:
BSP-116: Invalid authn test fix.

0x103 committed:
BSP-116: authorize and token tests updated.

podefr commented on src/commands/authn/authnCommand.ts:
```diff
@@ -0,0 +1,24 @@
+    readonly username: string;
+    readonly password: string;
+
+    constructor(data: unknown) {
```
👍

podefr commented on src/commands/authorize/authorizeCommand.ts:
```diff
@@ -0,0 +1,61 @@
+    client_id: string;
+    redirect_uri: string;
+    scope: string;
+    cookie?: string | undefined;
```
undefined is inferred by virtue of using an optional type so we can remove the `| undefined`

0x103 replied:
Yes, of course.

podefr commented on src/commands/token/exchangeTokenCommand.ts:
```diff
@@ -0,0 +1,47 @@
+    readonly client_id: string;
+    readonly client_secret: string;
+    readonly code: string;
+    readonly redirect_uri: string | undefined;
```
if `redirect_uri` is undefined we should mark it optional and remove the `| undefined`

0x103 replied:
Yes !

podefr commented on src/commands/token/exchangeTokenCommand.ts:
```diff
@@ -0,0 +1,47 @@
+    readonly code: string;
+    readonly redirect_uri: string | undefined;
+
+    constructor(data: unknown) {
```
should this be `data: ExchangeTokenInterface` ?

0x103 replied:
If we specify an interface we will have a problem in the deserialize function. The same goes if we specify a Record type, when we will call the function's constructor with the deserialized result, we will have a type error.

podefr commented on src/commands/token/refreshTokenCommand.ts:
```diff
@@ -0,0 +1,41 @@
+    grant_type: string;
+    refresh_token: string;
+    basic_auth: string;
+    scope?: string | undefined;
```
ditto

podefr commented on src/commands/token/refreshTokenCommand.ts:
```diff
@@ -0,0 +1,41 @@
+    readonly basic_auth: string;
+    readonly scope?: string | undefined;
+
+    constructor(data: unknown) {
```
same here, should that be `data: RefreshTokenInterface`?

podefr commented on src/commands/token/refreshTokenCommand.ts:
```diff
@@ -0,0 +1,41 @@
+    readonly grant_type: string;
+    readonly refresh_token: string;
+    readonly basic_auth: string;
+    readonly scope?: string | undefined;
```
ditto

podefr commented on src/commands/userinfo/userinfoCommand.ts:
```diff
@@ -0,0 +1,17 @@
+export default class UserinfoCommand implements UserinfoInterface {
+    readonly bearerToken: string;
+
+    constructor(data: unknown) {
```
`data: UserinfoInterface`

podefr commented on src/config.ts:
```diff
@@ -0,0 +1,74 @@
+    SESSIONS_ENDPOINT = ""
+} = process.env;
+
+const HANDLERS_SUBJECTS = {
```
this should be an enum

```
enum HANDLERS_SUBJECTS {
    TOKEN: "Token",
...
}
```

https://www.typescriptlang.org/docs/handbook/enums.html

tkjwa commented on src/commands/authorize/authorizeCommand.ts:
```diff
@@ -0,0 +1,61 @@
+    readonly state: string;
+    readonly client_id: string;
+    readonly redirect_uri: string;
+    readonly scope: string;
```
Are there restriction on scopes that we should check in the constructor ? what happens if someone call the endpoint with a "toto" scope ?

0x103 replied:
If someone call the endpoint with an invalid scope, an error is returned.

tkjwa commented on src/commands/authorize/authorizeCommand.ts:
```diff
@@ -0,0 +1,61 @@
+    readonly redirect_uri: string;
+    readonly scope: string;
+    readonly cookie?: string | undefined;
+    readonly sessionToken?: string | undefined;
```
Should the session token match a given format (can we test it with a regex) ?

0x103 replied:
Currently, the session token always start with `20111` (but nothing explain what it means, so I don't know if this can change) and have a length of 55 chars.
We could test the token length.

tkjwa commented on src/commands/authorize/authorizeCommand.ts:
```diff
@@ -0,0 +1,61 @@
+    readonly cookie?: string | undefined;
+    readonly sessionToken?: string | undefined;
+
+    constructor(data: unknown) {
```
@jwa-lab/jwa-lab-core  As a general question, should we add more validation than just types in the constructors ? this could also require to add different errors.

tkjwa commented on src/commands/authn/authnCommand.ts:
```diff
@@ -0,0 +1,24 @@
+        const { username, password } = data as AuthnInterface;
+
+        if (typeof username !== "string") {
+            throw new Error("INVALID_USERNAME");
```
Is it enough as an error message or should we have something like: `username must be a string, integer given (value: 123)`
Will this error bubble up to the user (the api client using our services) or is it only internal ?

0x103 replied:
We could have something more complex, but if we do I think we should create our own library then.
This error message if thrown back to the client, yes.

tkjwa commented on src/commands/token/exchangeTokenCommand.ts:
```diff
@@ -0,0 +1,47 @@
+        const { grant_type, client_id, client_secret, code, redirect_uri } =
+            data as ExchangeTokenInterface;
+
+        if (typeof grant_type !== "string") {
```
We should check also the value as the accepted ones are restricted

0x103 replied:
We can do it also here, yes.

tkjwa commented on tests/authorize.test.js:
```diff
@@ -0,0 +1,54 @@
+            expect(jsonCodec.decode(authorizeResponse.data)).toEqual(
+                expect.objectContaining({
+                    code: expect.any(String),
+                    redirect_uri: expect.any(String)
```
Shouldn't it be `http://localhost:3000/auth/callback` instead of any string ?

0x103 replied:
It could be, but I preferred to only check a string as I don't know if this callback url will stay forever in the used Okta app.

tkjwa reviewed (approved):
a few generic questions left

podefr commented on src/private/authnPrivateHandler.ts:
```diff
@@ -0,0 +1,70 @@
+                    const authnCommand = new AuthnCommand(data);
+                    const authnResponse = (await authnService.authn(
+                        authnCommand
+                    )) as AuthnResponseInterface;
```
I think `authnService.authn()` should return a Promise of `AuthnResponseInterface` instead of something generic that we need to type cast. `ConnectorResponse` should be a type used internally in the services

0x103 replied:
Indeed yes !

podefr commented on src/nats/nats.ts:
```diff
@@ -0,0 +1,109 @@
+import {
```
this is a kind of service, we should move it to the services folder

podefr commented on src/services/oktaSessionService.ts:
```diff
@@ -0,0 +1,34 @@
+    ): Promise<ConnectorResponse> {
+        const { sessionToken } = createSessionCommand;
+        const payload = {
+            sessionToken: sessionToken
```
you can use the shorthand notation here:

```
const payload = {
    sessionToken
}
```

podefr commented on src/services/oktaSessionService.ts:
```diff
@@ -0,0 +1,34 @@
+        this.restConnector = restConnector;
+    }
+
+    public async createSession(
```
most of the services methods return a `Promise< ConnectorResponse>` but ConnectorResponse is an interface used between the service and the restConnector. If we return a ConnectorResponse, the consumer of this methd will have to "guess" the type of the returned object. Let's do a typecast here from ConnectorResponse to `CreateSessionResponse`

podefr replied:
could you also make the change in all other services?

0x103 replied:
Yes !

0x103 committed:
BSP-116: Queues, Services responses, Commands params types & README

0x103 committed:
BSP-116: Prettier.

0x103 committed:
BSP-116: prettier now ignore the README.md

0x103 committed:
BSP-116: Authorize error test case fixed with new error message.

0x103 committed:
BSP-116; CircleCI Machine & Orb bump.

0x103 committed:
BSP-116: Back to npm install for lint step instead of npm ci.

0x103 committed:
BSP-116: CircleCI Test.

0x103 committed:
BSP-116: CircleCI Test.

0x103 committed:
BSP-116: CircleCI node back to default

0x103 committed:
BSP-116: CI lint executor

0x103 committed:
BSP-116 Stringified CI tag for lint.

0x103 committed:
BSP-116: Unsupported node 14 -> node 14.17

0x103 committed:
BSP-166: Using npm install instead of ci because of fsevents not being available.

0x103 committed:
BSP-116: npm < 7 on cimg 14, trying cimg16.6

0x103 committed:
BSP-166: CI lint step fixed & updated, checking test step

0x103 committed:
BSP-116: CI test step OK (nvm 14.7.3 & npm v7.20.3)

0x103 committed:
BSP-116: Prettier.

0x103 committed:
BSP-116: Types & Constructors fixes & commands moved to single folder

0x103 committed:
BSP-116: refreshTokenCommand fixed.

podefr committed:
BSP-116: Authentication Service

- Make commands constructor stricter
- Update deserialize function
- Make ConnectorResponse generic
- Fix issue with Refreshing Token
- Update documentation
- Update airlock

0x103 commented:
Commit squashed
